# GRIEVOUS

## Generalized Realignment of Innocuous and Essential Variants Otherwise Utilized as Skewed

<p align="center">
<img src="imgs/GRIEVOUS_Logos/GRIEVOUS Red Blade Logo.png" width = 350>
<img src="imgs/GRIEVOUS_Logos/GRIEVOUS Double Blades.png" width = 350>
</p>

*Hello There!* Welcome to GRIEVOUS, your command-line general to order the variant galaxy. In these hallowed archives, you will find all-manners of information for releasing GRIEVOUS on the intractable, inimical, insidious problem of variant harmonization in genomic datasets. Sound like a load of waffle? Okay, I'll attempt to eloquently expound...

Whether conducting a genome-wide association study (GWAS) or formulating, validating, and/or deploying a polygenic risk score (PRS), cross-dataset variant consistency in indexing and orientation is integral to the fidelity of your results. Previously the unenviable onus of variant standardization and allele consistency across all datasets of interest, fell on you, an esteemed, reputable and rambunctious researcher. GRIEVOUS is your command-line tool for guaranteeing cross-dataset variant consistency, freeing you from the shackles of time-consuming (and error-prone) tedium. Specifically, GRIEVOUS homogenizes indices across variants and reorients all biallelic SNPs across all datasets of interest (regardless of the number) consistently, handling common-issues such as duplications or multiple-indexing. GRIEVOUS can also be used to extract the set of biallelic SNPs common across all datasets of interest. Curious as to what types of data upon which you can unleash GRIEVOUS? Outstanding, I love the inquisitiveness (you must be a stellar scientist)! The answer would be summary statistics and genotype files.

Still lost? *Sigh,* perhaps a more thorough elaboration is required. Okay then, for a comprehensive detailed dive into GRIEVOUS, jump on over to our [documentation and tutorial](https://github.com/jvtalwar/GRIEVOUS/wiki)!  

Your time is valuable, and your research invaluable. Be sure to reclaim your time and guarantee the integrity of your results with GRIEVOUS. If you do, don't forget to cite our paper (link coming soon)! Until we meet again in another repository my fellow gentleperson of erudition, I wish you good fortune in the analyses to come (...*though in my experience, there's no such thing as luck*).

*[JVT](https://jvtalwar.github.io/dk-isle/)*  

***

 *Overwhelmed by the idea of consulting classical software documentation?* Perhaps you'd be interested in your very own GRIEVOUS artificial intelligence consultant (or as I like to call it, droideka). If so, shoot me a message and I'll add you to the waitlist for **DRAGONBREATH**, my (soon to be released) personalized LLM helper for bioinformatics software and data. Until it's release though, I guess you're stuck with the documentation. Fear not! I'll do all I can to combat the soporific and somniferous documentation connotation and ensure your conversion to the dark side of the force. 
